File: 1주차/홍종화/2792.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

class Main {
    static int[] jewelry;
    static int children, color;

    public static void main(String[] argv) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] line = br.readLine().split(" ");
        children = Integer.parseInt(line[0]);
        color = Integer.parseInt(line[1]);

        jewelry = new int[color];
        int max = 0;
        for (int i = 0; i < color; i++) {
            jewelry[i] = Integer.parseInt(br.readLine());
            if (max < jewelry[i]) {
                max = jewelry[i];
            }
        }
        int l = 1;
        int r = max; // [1, 10^9]
        while (l <= r) {
            int mid = (l + r) / 2;
            if (isPossible(mid)) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        System.out.println(l);
    }

    public static boolean isPossible(int mid) {
        int cnt = 0;
        for (int i = 0; i < color; i++) {
            cnt += Math.ceil(jewelry[i] / (double) mid);
            if (children < cnt) {
                return true;
            }
        }
        return false;
    }
}

File: 1주차/손상렬/2792.cpp
#include <cstdio>
#include <algorithm>
using namespace std;
const int SIZE = 300'000;
const int INF = 1e9;
int N, M, color[SIZE];
bool isPossible(int m){
    int cnt = 0;
    for (int i = 0; i < M; i++){
        cnt += (color[i] - 1) / m + 1;
    }
    return cnt > N;
}
int main(){
    scanf("%d %d", &N, &M);
    for (int i = 0; i < M; i++) scanf("%d", color + i);
    int lo = 0, hi = INF;
    while (lo + 1 < hi){
        int mid = (lo + hi) / 2;
        if (isPossible(mid)) lo = mid;
        else hi = mid;
    }
    printf("%d\n", hi);
}